### PR TITLE
[3] add a dns_module option that allows overriding the DNS resolver (useful for testing)

### LINKED ISSUE #3: Don't access network in test environment
Resolving DNS in test environment may make tests flaky. An example from CI build:
```elixir
** (Socket.Error) timeout
    (socket 0.3.13) lib/socket/datagram.ex:46: Socket.Datagram.recv!/2
    (dns 2.3.0) lib/dns.ex:76: DNS.query/4
    (dns 2.3.0) lib/dns.ex:23: DNS.resolve/4
    (safeurl 0.2.0) lib/safeurl.ex:267: SafeURL.resolve_address/1
    (safeurl 0.2.0) lib/safeurl.ex:136: SafeURL.allowed?/2
    (safeurl 0.2.0) lib/safeurl.ex:177: SafeURL.validate/2
```

### PR DESCRIPTION
**Pull request by bhan-slab**

### CODE REVIEW

**Comment by sheharyarn** [comment]
I changed the base of this PR to the branch in #5 to simplify the git diff. Please merge #5 before this PR.

**Review by sheharyarn — CHANGES_REQUESTED** [review]
Mocking DNS in our own tests would be the wrong strategy. The goal in #3 is to disable DNS resolution when needed, which can be useful to other Elixir projects that use SafeURL.

My suggestion is to introduce a `resolve_dns` [config](https://github.com/slab/safeurl-elixir/blob/bc9f1ed421d4e3a6b6ee47ccd333e1f0728b13c3/mix.exs#L46-L53) that defaults to `true`. We should [resolve DNS](https://github.com/slab/safeurl-elixir/blob/96d4e672857521fd43123e314e855756d8ed3ec1/lib/safeurl.ex#L267-L270) only when this config is true.

**Comment by bhan-slab** [comment]
> Mocking DNS in our own tests would be the wrong strategy. The goal in #3 is to disable DNS resolution when needed, which can be useful to other Elixir projects that use SafeURL.
> 
> My suggestion is to introduce a `resolve_dns` [config](https://github.com/slab/safeurl-elixir/blob/bc9f1ed421d4e3a6b6ee47ccd333e1f0728b13c3/mix.exs#L46-L53) that defaults to `true`. We should [resolve DNS](https://github.com/slab/safeurl-elixir/blob/96d4e672857521fd43123e314e855756d8ed3ec1/lib/safeurl.ex#L267-L270) only when this config is true.

what should the downstream behavior be when resolving DNS is turned off? the downstream logic relies on getting an IP address.

**Comment by martosaur** [comment]
I think the biggest drawback of simply mocking calls to `DNS` module is that it won't work for other people who use this library.

@sheharyarn should we go with `DNSResolver` behavior pattern so that users could swap it for mock module in tests? If we make a config option to skip dns resolution people won't actually be able to test both positive and negative scenarios?

**Comment by sheharyarn** [comment]
That's a really good point. My thinking is we don't want to prematurely optimize or build solutions to problems that have not been reported enough yet. But that said, simply plugging in a module in the config is a good solution for library users that offers a lot of flexibility.

We can just implement the `SafeURL.DNSResolver` behaviour with one `resolve(domain)` callback as you suggested, but leave out the behaviour implementation wrapper around the `DNS` module and directly use it in the config as `dns_resolver: DNS`.

**Comment by bhan-slab** [comment]
@sheharyarn @martosaur thanks for the suggestions! i've implemented an approach using `behaviour`. could you take a look?

File: mix.lock [hunk 1]
```diff
@@ -10,8 +10,10 @@
   "makeup": {:hex, :makeup, "1.0.5", "d5a830bc42c9800ce07dd97fa94669dfb93d3bf5fcf6ea7a0c67b2e0e4a7f26c", [:mix], [{:nimble_parsec, "~> 0.5 or ~> 1.0", [hex: :nimble_parsec, repo: "hexpm", optional: false]}], "hexpm", "cfa158c02d3f5c0c665d0af11512fed3fba0144cf1aadee0f2ce17747fba2ca9"},
   "makeup_elixir": {:hex, :makeup_elixir, "0.15.2", "dc72dfe17eb240552857465cc00cce390960d9a0c055c4ccd38b70629227e97c", [:mix], [{:makeup, "~> 1.0", [hex: :makeup, repo: "hexpm", optional: false]}, {:nimble_parsec, "~> 1.1", [hex: :nimble_parsec, repo: "hexpm", optional: false]}], "hexpm", "fd23ae48d09b32eff49d4ced2b43c9f086d402ee4fd4fcb2d7fad97fa8823e75"},
   "makeup_erlang": {:hex, :makeup_erlang, "0.1.1", "3fcb7f09eb9d98dc4d208f49cc955a34218fc41ff6b84df7c75b3e6e533cc65f", [:mix], [{:makeup, "~> 1.0", [hex: :makeup, repo: "hexpm", optional: false]}], "hexpm", "174d0809e98a4ef0b3309256cbf97101c6ec01c4ab0b23e926a9e17df2077cbb"},
+  "meck": {:hex, :meck, "0.9.2", "85ccbab053f1db86c7ca240e9fc718170ee5bda03810a6292b5306bf31bae5f5", [:rebar3], [], "hexpm", "81344f561357dc40a8344afa53767c32669153355b626ea9fcbc8da6b3045826"},
```
Comment (martosaur) [inline mix.lock]: these are not needed anymore?

**Review by martosaur — APPROVED** [review]
Looks good! 🎉

File: mix.lock (hunk 1 above)
Comment (bhan-slab) [reply mix.lock]: removed, thanks!

**Pull request merged**